diff --git a/ldo.c b/ldo.c
index ea052950..0d6ebc6c 100644
--- a/ldo.c
+++ b/ldo.c
@@ -10,7 +10,6 @@
 #include "lprefix.h"
 
 
-#include <setjmp.h>
 #include <stdlib.h>
 #include <string.h>
 
@@ -62,6 +61,7 @@
 
 #elif defined(LUA_USE_POSIX)				/* }{ */
 
+#include <setjmp.h>
 /* in POSIX, try _longjmp/_setjmp (more efficient) */
 #define LUAI_THROW(L,c)		_longjmp((c)->b, 1)
 #define LUAI_TRY(L,c,a)		if (_setjmp((c)->b) == 0) { a }
@@ -69,6 +69,7 @@
 
 #else							/* }{ */
 
+#include <setjmp.h>
 /* ISO C handling with long jumps */
 #define LUAI_THROW(L,c)		longjmp((c)->b, 1)
 #define LUAI_TRY(L,c,a)		if (setjmp((c)->b) == 0) { a }
diff --git a/liolib.c b/liolib.c
index c5075f3e..56c0e733 100644
--- a/liolib.c
+++ b/liolib.c
@@ -299,6 +299,10 @@ static int io_popen (lua_State *L) {
   return (p->f == NULL) ? luaL_fileresult(L, 0, filename) : 1;
 }
 
+#if defined(LUA_TMPFILE_MISSING)
+FILE * my_tmpfile();
+#define tmpfile my_tmpfile
+#endif
 
 static int io_tmpfile (lua_State *L) {
   LStream *p = newfile(L);
diff --git a/loslib.c b/loslib.c
index ba80d72c..b1bb95a6 100644
--- a/loslib.c
+++ b/loslib.c
@@ -101,7 +101,40 @@
 */
 #if !defined(lua_tmpnam)	/* { */
 
-#if defined(LUA_USE_POSIX)	/* { */
+
+#if defined(LUA_TMPFILE_MISSING)
+
+#define LUA_TMPNAMBUFSIZE	32
+#if !defined(LUA_TMPNAMTEMPLATE)
+#define LUA_TMPNAMTEMPLATE	"/tmp/lua_XXXXXX"
+#endif
+
+FILE * findtempfile(char* fn, int n) {
+	for (int i = 0; i < 99999; ++i) {
+		snprintf(fn, n, "%s%d", LUA_TMPNAMTEMPLATE, i);
+		FILE * f = fopen(fn, "r");
+		if (!f) return f;
+		fclose(f);
+	}
+	return NULL;
+}
+
+FILE * my_tmpfile() {
+	char fn[LUA_TMPNAMBUFSIZE];
+	return findtempfile(fn, sizeof(fn));
+}
+
+#define lua_tmpnam(fn, e) {\
+	FILE * f = findtempfile(fn, sizeof(fn));\
+	if (f) {\
+		fclose(f);\
+		e=0;\
+	} else {\
+		e=1;\
+	}\
+}
+
+#elif defined(LUA_USE_POSIX)	/* { */
 
 #include <unistd.h>
 
@@ -139,6 +172,7 @@
 #endif
 
 
+
 static int os_execute (lua_State *L) {
   const char *cmd = luaL_optstring(L, 1, NULL);
   int stat;
diff --git a/luaconf.h b/luaconf.h
index 33bb580d..1e465c8e 100644
--- a/luaconf.h
+++ b/luaconf.h
@@ -284,6 +284,7 @@
 ** the libraries, you may want to use the following definition (define
 ** LUA_BUILD_AS_DLL to get it).
 */
+#if !defined(LUA_API)
 #if defined(LUA_BUILD_AS_DLL)	/* { */
 
 #if defined(LUA_CORE) || defined(LUA_LIB)	/* { */
@@ -297,6 +298,7 @@
 #define LUA_API		extern
 
 #endif				/* } */
+#endif
 
 
 /*
