diff --git a/linit.c b/linit.c
index 69808f84..6376f2d9 100644
--- a/linit.c
+++ b/linit.c
@@ -34,6 +34,9 @@
 #include "lualib.h"
 #include "lauxlib.h"
 
+// Chris: wedging this in there ...
+//#include "../luaffifb/ffi.h"
+int luaopen_ffi(lua_State* L);
 
 /*
 ** these libs are loaded by lua.c and are readily available to any Lua
@@ -50,6 +53,11 @@ static const luaL_Reg loadedlibs[] = {
   {LUA_MATHLIBNAME, luaopen_math},
   {LUA_UTF8LIBNAME, luaopen_utf8},
   {LUA_DBLIBNAME, luaopen_debug},
+
+// Chris: wedging this in there ...
+// I guess I should insert it into one of the loaders instead of always here , but meh, this is a giant hack anyways
+  {"ffi", luaopen_ffi},
+
   {NULL, NULL}
 };
 
diff --git a/luaconf.h b/luaconf.h
index 33bb580d..71208dc5 100644
--- a/luaconf.h
+++ b/luaconf.h
@@ -284,6 +284,7 @@
 ** the libraries, you may want to use the following definition (define
 ** LUA_BUILD_AS_DLL to get it).
 */
+#if !defined(LUA_API)
 #if defined(LUA_BUILD_AS_DLL)	/* { */
 
 #if defined(LUA_CORE) || defined(LUA_LIB)	/* { */
@@ -292,11 +293,14 @@
 #define LUA_API __declspec(dllimport)
 #endif						/* } */
 
+#elif defined __GNUC__
+# define LUA_API __attribute__((visibility("default")))
 #else				/* }{ */
 
 #define LUA_API		extern
 
 #endif				/* } */
+#endif
 
 
 /*
